### PR TITLE
Issue #248: Refactored handlers

### LINKED ISSUE #248: refactor handler
https://github.com/dotkernel/api/blob/4.0/src/User/src/Handler/UserHandler.php#L58

1. how to process the handler 
2. how to threat errors in a smart way

### PR DESCRIPTION
**Pull request by alexmerlin**
This PR does the following:
- simlify handler actions by removing responsability of handling exceptions
- enhances handler response status codes based on the response / exception
- handlers return HTTP status code based on response / exception
- creates/uses case-specific exceptions instead of the generic ones
- updates method docblocks with the correct exceptions thrown
- capture code-level exception messages, log them to the daily log and rethrow the same type of exception with a safe message that can be shown in a UI
- fixes user ResetPassword retrieval based on the `hash` from the path (vs always using the latest one)
- minor code cleanups
- replaced UUIDv1 usage with UUIDv4
- restored `#[ORM\HasLifecycleCallbacks]` attribute to all entities extending our `AbstractEntity` because without that, the `updatedAt` field does not get populated
- breaks ResponseTrait into HandlerTrait and ResponseTrait so that we can access `*Response` methods without including the `handle` method in a class
- return `204 No Content` on successful DELETE operations

---

Tasks remaining:

- [x] fix any issues reported by the reviewers
- [x] fix existing unit tests
- [ ] add more unit tests

### CODE REVIEW

File: src/Admin/src/Handler/AdminAccountHandler.php [hunk 1]
```diff
@@ -24,39 +26,37 @@ class AdminAccountHandler implements RequestHandlerInterface
...
+     * @throws ConflictException
+     * @throws FormValidationException
+     * @throws NotFoundException
+     */
     public function patch(ServerRequestInterface $request): ResponseInterface
     {
         $inputFilter = (new UpdateAdminInputFilter())->setData($request->getParsedBody());
         if (! $inputFilter->isValid()) {
-            return $this->errorResponse($inputFilter->getMessages());
+            throw (new FormValidationException())->setMessages($inputFilter->getMessages());
```
Comment (marioradu05) [inline src/Admin/src/Handler/AdminAccountHandler.php]: What if we initialize the `FormValidationException` by passing the messages as construtor parameters?
`throw (new FormValidationException($inputFilter->getMessages()))`

File: src/User/src/EventListener/UserAvatarEventListener.php [hunk 2]
```diff
@@ -47,7 +47,7 @@ private function setAvatarUrl(UserAvatar $avatar): void
             sprintf(
                 '%s/%s/%s',
                 rtrim($this->config['uploads']['user']['url'], '/'),
-                $avatar->getUser()->getUuid()->toString(),
+                $avatar->getUser()->getUuid()?->toString(),
```
Comment (marioradu05) [inline src/User/src/EventListener/UserAvatarEventListener.php]: We should try to remove the nullable return type from `getUuid()` declaration.

**Review by marioradu05 — CHANGES_REQUESTED** [review]

File: src/User/src/EventListener/UserAvatarEventListener.php (hunk 2 above)
Comment (alexmerlin) [reply src/User/src/EventListener/UserAvatarEventListener.php]: *entity->getUuid() will alwas return a `UuidInterface` - remove null-checks from all occurrences of `getUuid()?->toString()`.

File: src/Admin/src/Handler/AdminAccountHandler.php (hunk 1 above)
Comment (alexmerlin) [reply src/Admin/src/Handler/AdminAccountHandler.php]: `$inputFilter->getMessages()` returns a multidimensional associative array.
Unless serializing the errors somehow, I cannot use them to initialize `FormValidationException`'s parent class `Exception`.

That's why I opted for this approach of:
- initializing an empty `FormValidationException`
- setting messages as an array which is later used in `src/App/src/Handler/ResponseTrait.php` when catching `FormValidationException` and returning `$this->errorResponse($exception->getMessages())`.

**Comment by alexmerlin** [comment]
@MarioRadu @cPintiuta When you have time, take another look at this PR and leave your feedback with any observations you have

**Review by arhimede — CHANGES_REQUESTED** [review]
add 204 to DELETE 
201 when a resource is uploaded for the first time

**Comment by alexmerlin** [comment]
> add 204 to DELETE 201 when a resource is uploaded for the first time

Implemented.

**Review by arhimede — APPROVED** [review]

**Review by marioradu05 — APPROVED** [review]

**Pull request merged**